camel/bots/discord/discord_app.py

```
@@ -0,0 +1,375 @@
+            guild_id=guild_id
+        )
+        if not installation:
+            logger.error(f"No installation found for guild: {guild_id}")
+            return None
+
+        if (
+            installation.token_expires_at
+            and datetime.now() >= installation.token_expires_at
+        ):
+            logger.info(
+                f"Access token expired for guild: {guild_id}, "
+                f"refreshing token..."
+            )
+            new_access_token = await self.refresh_access_token(
+                installation.refresh_token
+            )
+            if new_access_token:
+                installation.access_token = new_access_token
+                installation.token_expires_at = datetime.now() + timedelta(
+                    seconds=3600
+                )
+                await self.installation_store.save(installation)
+                return new_access_token
+            else:
+                logger.error(
+                    f"Failed to refresh access token for guild: {guild_id}"
+                )
+                return None
+
+        return installation.access_token
+
+    async def save_installation(
+        self,
+        guild_id: str,
+        access_token: str,
+        refresh_token: str,
+        expires_in: int,
+    ):
+        r"""Save the installation information for a given guild.
+
+        Args:
+            guild_id (str): The ID of the guild where the bot is installed.
+            access_token (str): The access token for the guild.
+            refresh_token (str): The refresh token for the guild.
+            expires_in: (int): The expiration time of the
+                access token.
+        """
+        if not self.oauth_flow:
+            logger.warning(
+                "OAuth is not enabled. Missing client_id, "
+                "client_secret, or redirect_uri."
+            )
+            return None
+        assert self.installation_store is not None
```

It would be better to raise an error here.